reconcile/gitlab_owners.py

```diff
@@ -0,0 +1,233 @@
+            # Each change that was not yet approved will generate
+            # a report message
+            if not change_approved:
+                report[change_path] = (f'one of {change_owners} needs to '
+                                       f'approve the change')
+
+        # Empty report means that all changes are approved
+        if not report:
+            approval_status['approved'] = True
+            return approval_status
+
+        # Since we have a report, let's check if that report was already
+        # used for a comment
+        comments = self.gitlab.get_merge_request_comments(self.mr_id)
+        for comment in comments:
+            # Only interested on our own comments
+            if comment['username'] != 'devtools-bot':
+                continue
+
+            # Removing the pre-formatted markdown from the comment
+            json_body = comment['body'].lstrip('```\n').rstrip('\n```')
+
+            try:
+                body = json.loads(json_body)
+                # If we find a comment equals to the report,
+                # we don't return the report
+                if body == report:
+                    return approval_status
+            except json.decoder.JSONDecodeError:
+                continue
+
+        # At this point, the MR was not approved and the report
+        # will be used for creating a comment in the MR.
+        json_report = json.dumps(report, indent=4)
+        markdown_json_report = f'```\n{json_report}\n```'
+        approval_status['report'] = markdown_json_report
+        return approval_status
+
+    def has_approval_label(self):
+        labels = self.gitlab.get_merge_request_labels(self.mr_id)
+        return APPROVAL_LABEL in labels
+
+
+def run(dry_run=False):
+    apps = queries.get_apps_gitlab_owners_repos()
+
+    # Collecting the repositories with gitlabOwners=True
+    # in App-Interface
+    gitlab_owners_repos = []
+    for app in apps:
+        for code_components in app.values():
+            if code_components is None:
+                continue
+            for code_component in code_components:
+                if not code_component['gitlabOwners']:
+                    continue
+                gitlab_owners_repos.append(code_component['url'])
+
+    for repo in gitlab_owners_repos:
+        instance = queries.get_gitlab_instance()
+        settings = queries.get_app_interface_settings()
+        gitlab_cli = GitLabApi(instance, project_url=repo,
+                               settings=settings)
+
+        project_owners = ProjectOwners(gitlab_client=gitlab_cli)
+
+        for mr in gitlab_cli.get_merge_requests(state=MRState.OPENED):
+            mr_approval = MRApproval(gitlab_client=gitlab_cli,
+                                     merge_request_id=mr.iid,
+                                     project_owners=project_owners,
+                                     dry_run=dry_run)
+
+            if mr_approval.has_approval_label():
```

Comment:
what happens if the MR is updated to include unauthorized changes after the `approved` label is added?

Reply:
fixed! Now we restart the approval process after each new push